third_party/move/move-compiler-v2/src/pipeline/flush_writes_processor.rs

```diff
@@ -82,52 +194,87 @@ impl FunctionTargetProcessor for FlushWritesProcessor {
+    /// In the basic block given by `block_range` part of `code`, extract the writes
+    /// to temporaries that are better flushed right away. At the offset where the
+    /// write happens, such temporaries are included, in the out param `flush_writes`.
+    fn extract_flush_writes_in_block(
+        block_range: RangeInclusive<CodeOffset>,
+        code: &[Bytecode],
+        use_def_links: &UseDefLinks,
+        flush_writes: &mut BTreeMap<CodeOffset, BTreeSet<TempIndex>>,
+    ) {
+        let upper = *block_range.end();
+        for offset in block_range {
+            let instr = &code[offset as usize];
+            // Only `Load` and `Call` instructions push temps to the stack.
+            // We need to find if any of these temps are better flushed right away.
+            if matches!(instr, Bytecode::Load(..) | Bytecode::Call(..)) {
+                for (dest_index, dest) in instr.dests().into_iter().enumerate() {
+                    let def = DefOrUsePoint {
+                        offset,
+                        index: dest_index,
+                    };
+                    if Self::could_flush_right_away(def, upper, use_def_links) {
+                        flush_writes.entry(offset).or_default().insert(dest);
                     }
                 }
             }
         }
     }
-}
 
-/// Is `temp` unused in `relevant_range`, based on `live_info` at the code offset
-/// where `temp` was written to?
-fn is_unused_in_block(
-    temp: TempIndex,
-    relevant_range: RangeInclusive<CodeOffset>,
-    live_info: &LiveVarInfoAtCodeOffset,
-) -> bool {
-    if let Some(info) = live_info.after.get(&temp) {
-        // Note: loop-carried uses are not considered here.
-        let all_usages_are_outside_block = info
-            .usage_offsets()
-            .iter()
-            .all(|usage| *usage <= *relevant_range.start() || *usage > *relevant_range.end());
-        all_usages_are_outside_block
-    } else {
-        // `temp` has no usages after `offset`.
-        true
+    /// Is the `def` better flushed right away?
+    /// `block_end` is the end of the block that has `def`.
+    fn could_flush_right_away(
+        def: DefOrUsePoint,
+        block_end: CodeOffset,
+        use_def_links: &UseDefLinks,
+    ) -> bool {
+        use_def_links.def_to_use.get(&def).map_or(true, |uses| {
+            let exactly_one_use = uses.len() == 1;
+            if !exactly_one_use {
+                // If there is more than one use, flush right away.
+                return true;
+            }
+            let use_ = uses.first().expect("there is exactly one use");
+            let use_outside_block = use_.offset <= def.offset || use_.offset > block_end;
+            if use_outside_block {
+                // If used outside the basic block, flush right away.
+                return true;
+            }
+            // If has intervening definition, flush right away.
+            Self::has_intervening_def(def, use_, use_def_links)
```

Comment:
No, Def/NextUse would not be sufficient for a few reasons.

1. We need to see if there are multiple or only out of block uses (the checks above) - in these cases, we have to flush the definition.
2. We need to see in a usage instruction, no previous operands (sources) have definitions in between. To get this, a "NextUse" of those definitions are not always well defined: you can have multiple non-dominating next uses.

Now, can we do this by computing what we need later from live vars directly? Yes, but the logic can get a bit complicated to reason about, and it I don't think it is easy to conclude that it is "much more efficient".

Def Use chains are also more general purpose, and as we extend this pass to potentially include more cases (like we have done in this PR), it would make it easier to extend.